test/unit/bucket/test_bucket.py

```diff
@@ -1729,6 +1731,80 @@ def _upload_part(self, large_file_id, part_number, part_data):
         )
 
 
+class TestBucketRaisingSession(TestUpload):
+    def get_api(self):
+        class B2SessionRaising(B2Session):
+            def __init__(self, *args, **kwargs):
+                self._raise_count = 0
+                self._raise_until = 1
+                super().__init__(*args, **kwargs)
+
+            def upload_part(
+                self,
+                file_id,
+                part_number,
+                content_length,
+                sha1_sum,
+                input_stream,
+                server_side_encryption=None
+            ):
+                if self._raise_count < self._raise_until:
+                    self._raise_count += 1
+                    raise B2ConnectionError()
+                return super().upload_part(
+                    file_id, part_number, content_length, sha1_sum, input_stream,
+                    server_side_encryption
+                )
+
+        class B2ApiPatched(B2Api):
+            SESSION_CLASS = staticmethod(B2SessionRaising)
+
+        self.api = B2ApiPatched(
+            self.account_info,
+            cache=self.CACHE_CLASS(),
+            api_config=B2HttpApiConfig(_raw_api_class=self.RAW_SIMULATOR_CLASS),
+        )
+        return self.api
+
+    def test_upload_chunk_retry_stream_open(self):
+        assert self.api.session._raise_count == 0
+        data = self._make_data(self.simulator.MIN_PART_SIZE * 3)
+        self.bucket.upload_unbound_stream(io.BytesIO(data), 'file1')
+        self._check_file_contents('file1', data)
+        assert self.api.session._raise_count == 1
+
+    def test_upload_chunk_stream_guard_closes(self):
+        data = self._make_data(self.simulator.MIN_PART_SIZE * 3)
+        large_file_upload_state = mock.MagicMock()
+        large_file_upload_state.has_error.return_value = False
+
+        class TrackedUploadSourceBytes(UploadSourceBytes):
```

Comment:
Have you considered using `MagicMock` instead of implementing it? Would it get you the thing you're looking for?

Reply:
copy-paste from DM in case anyone else is wondering about this:

> Using mocks was my initial approach, but I still needed the classes to make the correct things with data and interacting with other classes and it was getting more and more dirty...